Nb feat/list providers endpoint

nicobarra opened the pull request:
* Initialized and organized structure of the repo
* Created storage for providers, with its method that interacts with the DB
* Created struct and interface (this for next features) for Provider type
* Created route and handler for ListProviders endpoint
* Created main file that initializes storage and executes the api (fiber)

This endoint works, here is an image of an execution through insomnia:
<img width="819" alt="image" src="https://user-images.githubusercontent.com/71539596/198727849-3fc99309-d118-41a8-95f9-46de2d4a124f.png">

mtavano commented on internal/api/jobs/list-jobs-handler.go:
```diff
@@ -0,0 +1,26 @@
+package jobsapi
+
+import (
+	jobstorage "github.com/darchlabs/jobs/internal/storage/job"
```
may we can use only `storage.Job` which is better than `JS` . we should avoid those micro names based on its initials. since the package + name can give a better context of what is, we also can avoid rename imports so in that case we can better track the code at its usage

nicobarra replied:
Yes, I totally agree. I will change it.

mtavano commented on internal/api/jobs/route.go:
```diff
@@ -0,0 +1,56 @@
+type HandlerRes struct {
+	Payload    interface{}
+	HttpStatus int
+	err        error
```
why this is the only property that is not exported of `HandlerRes` ?

nicobarra replied:
Because I missed it. I'll fix

mtavano commented on internal/api/jobs/route.go:
```diff
@@ -0,0 +1,56 @@
+		payload, statusCode, err := handlerRes.Payload, handlerRes.HttpStatus, handlerRes.err
+		if err != nil {
+			return c.Status(fiber.StatusConflict).JSON(api.Response{
+				Error: err,
+				Data:  payload,
```
Why is this response using algo `Data` if there is an error?  are we planning to use `Data` + `Error` in case of error? maybe we should use `Meta` instead? 🤔 

Im not sure about the idea behind the response but I think that Error should be the only property when we want to expose information regarding that

nicobarra replied:
Ok, I thought that could be useful but I will only use Meta like you suggest

mtavano commented on internal/api/providers/list-providers-handler.go:
```diff
@@ -0,0 +1,24 @@
+package providers
+
+import providerstorage "github.com/darchlabs/jobs/internal/storage/provider"
```
same about storage here. `storage.Provider` would be good enough to represent a job/provider/whatever storage

mtavano commented on internal/api/providers/route.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+// handler's response
+type HandlerRes struct {
+	Payload    interface{}
+	HttpStatus int
+	err        error
+}
```
we should not repeat this struct since is used in more than one place. Evaluate to return more than one element as `(interface{}, int, error)` instead add overhead to every handler package just to return things can be simply returned as different plain parameters

nicobarra replied:
Like putting this struct in `api.go` file that is shared in all the pkg?

mtavano commented on internal/api/providers/route.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+// Func that receives the returns from handlers and creates an http response
+func HandleFunc(fn handler, ctx Context) func(c *fiber.Ctx) error {
+	return func(c *fiber.Ctx) error {
+		// set headers
+		c.Accepts("application/json")
+
+		// Exec handler func and get its response
+		handlerRes := fn(ctx)
+		payload, statusCode, err := handlerRes.Payload, handlerRes.HttpStatus, handlerRes.err
+		if err != nil {
+			return c.Status(fiber.StatusConflict).JSON(api.Response{
+				Error: err,
+				Data:  payload,
+			})
+		}
+
+		// Prepare response
+		res := api.Response{
+			Meta: map[string]interface{}{"statusCode": statusCode},
+			Data: payload,
+		}
+
+		// Return response
+		return c.Status(fiber.StatusOK).JSON(res)
+	}
+}
```
I think that we are repeating tons of code that is way not needed. We should define these things into a top-level package and use them. since we have a common folder api, I think that this is better to go into api and then import handlers that are needed. In general, this kind of things are only useful for microservices but we are not thinking in microservices se we could talk about it for better project building and avoid these things

nicobarra replied:
Yes, but there is a slight but important difference between handler decorators func of the different `route.go` files.

And the difference is that one needs to receive params and the another doesn't. For that reason is that I used 2 very similar func's, but I agree with you that there is a lot of code repeated and it could be improved.

mtavano commented on internal/provider/provider.go:
```diff
@@ -0,0 +1,14 @@
+
+// Interface required for each new implementation
+type Create interface {
+	Setup(address string, abi string, checkMethod string, actionMethod string, checkType string, checkValue string) bool
```
for these things I think that is better to create a struct like Config, Input or things like that and keep consistency. we are using structs for things that are not needed and also omitting structs when is useful 💡

nicobarra replied:
Ok

mtavano commented on internal/storage/job/job.go:
```diff
@@ -0,0 +1,41 @@
+	"github.com/darchlabs/jobs/internal/storage"
+)
+
+type JS struct {
```
As I mentioned before, naming is important and the package path gives us context about what the package is for and its structs gives us the rest of the needed information

mtavano commented on internal/storage/provider/provider.go:
```diff
@@ -0,0 +1,42 @@
+	"github.com/darchlabs/jobs/internal/storage"
+)
+
+type PS struct {
```
same about naming

mtavano commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,27 @@
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+type S struct {
```
same about naming

mtavano reviewed (commented):
I know that I left a bunch of comments but is important to be consistent from the beggining of the project since the tech debt is very painful to tackle once the project is in an advance phase

nicobarra commented:
Finished list providers endpoint.

Changed leveldb storage implementation for a hardcoded struct in the handler, since that was decided by the team that was the better option because there won't be too much job providers implementation and they need to be added manually.
Also removed context from handler and route since it is not necessary.

This is the response in Postman:
<img width="673" alt="image" src="https://user-images.githubusercontent.com/71539596/199537570-694514cb-80c1-42bb-b89b-bf52e10560f3.png">
